rfcs/0007-handling-corruption.md

```diff
@@ -0,0 +1,392 @@
+affected by the corruption remain accessible, and new keys can be added to the
+database.
+
+For my research into RocksDB and Pebble, See  https://github.com/thrawn01/db-corruption
+
+## Goals
+The goal of this proposal is to develop a robust and flexible error reporting
+and corruption handling system. This system will be suitable for building both
+durability and availability systems on top of SlateDB.
+
+- Ensure the database remains available in the face of corruption.
+- Report corruption errors directly to the caller when corruption is detected.
+- Provide a corruption API that developers and operators can use to repair and
+  attempt recovery of corrupted data.
+- Enhance the robustness of the WAL implementation by attempting to repair and
+  recover as much of the WAL as possible, rather than truncating the remaining
+  entries.
+
+### Proposed Solution
+To provide support both Availability and Durability, I propose adding a callback
+function called `DBOptions.on_corruption` which is called when SlateDB encounters
+corruption during operation. The use of a call back allows developers to decide their
+desired behavior when corruption is detected. The return value of the callback
+then informs SlateDB of the users desire to abort the current operation with an
+error, or to continue operation attempting to skip the encountered corruption.
+
+To support developers who value availability, the `on_corruption` callback can
+be designed to log or notify the operator of the corruption, but otherwise
+remaining available.
+
+To support developers who value durability, the `on_corruption` callback can be
+designed to log or notify the operator of the corruption, then return an error,
+indicating the operators desire to make the database unavailable in order to 
+avoid losing data which may be recoverable.
+
+Additionally, the call back can be used to notify logical systems built on top
+of SlateDB that corruption exists and take appropriate action to repair and
+restore corrupted data from replicas or backups.
+
+```go
+  DBOptions.on_corruption = func(d CorruptionDetails) error {
+    log.Error("Corruption Detected: %s", d.String())
+    // Could page the operator
+    if err := pageOperator(d); err != nil {
+      log.Error("while paging operator: %s", err)
+    }
+
+    // Could inform the logical system to repair data
+    // via a replica
+    if err := signalCorruption(d); err != nil {
+      log.Error("while signaling: %s", err)
+    }
+
+    // If we want the database to become unavailable until 
+    // corruption is fixed, return an error and SlateDB will
+    // preform a safe shutdown.
+    return fmt.Errorf("Corruption Detected: %s", d.String())
+
+    // Else we return no error, and the database remains available.
+    return nil
+  }
+```
+
+If not defined by the user, the `DBOptions.on_corruption` callback should
+both log and return an error. This ensures that SlateDB is durable by default.
+
```

Comment:
At the file format level, WAL and compacted SSTs are the same. Why does the corruption RFC differentiate them?

Reply:
The way I think about this, corruption in the WAL is more serious than corruption in the SSTs. 

- In the case of SST corruption, it's likely the operator either 1. Has a backup 2. Has a logical system managing replicas it can use to recover the corruption.
- In the case of WAL corruption, it's likely the only copy of the data. 

Now that I'm typing this out, the `on_corruption` callback should indicate the source of the corruption, as the operator might want to continue operation on SST corruption, but require user intervention on WAL corruption. I'll add this to the RFC.

To your point, both inspect and repair operations for WAL and SST are likely to be nearly identical and share code. But semantically, I think they are separate.